### PR TITLE
Implementation of Lokta Volterra with NumPyro

### PR DESCRIPTION
**Pull request by EiffL**
This is an alternative to #1, for illustration purposes, not saying we have to use this one. 
But it shows how a numpyro model can be made more compact

### CODE REVIEW

**Comment by EiffL** [comment]
The interest of formatting things this way is that any model we can define as a numpyro function with the following assumptions regarding variable names:
- `theta` corresponds to model parameters to be inferred
- `z` corresponds to model latent parameters
- `y` corresponds to model output (sampled from likelihood)

Then for any model with these conventions, we can use a function like `sbids.tasks.get_samples_and_scores` to extract the quantities of interest.

**Comment by EiffL** [comment]
Also, this adds a [notebook](https://github.com/Justinezgh/SBI-Diff-Simulator/blob/u/EiffL/numpyro/notebooks/NumpyroLotkaVolterra.ipynb) that  demonstrates how to use the numpyro NUTS sampler, to get 10,000 samples in like 10mins (without even using parallel chains), and shows how to use ChainConsumer to make pretty plots of overlapping distributions.

image

File: sbids/tasks/utils.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        y_prob = model_trace['y']['fn'].log_prob(jax.lax.stop_gradient(model_trace['y']['value']))
+
+        sample = {'theta': model_trace['theta']['value'],
+                  'y': model_trace['y']['value'],
+                  'z': model_trace['z']['value']}
+        
+        if score_type == 'density':
+            return theta_prob.sum() + z_prob.sum() + y_prob.sum(), sample
+        elif score_type == 'conditional':
+            return y_prob.sum() + z_prob.sum(), sample
```
Comment (Justinezgh) [inline sbids/tasks/utils.py:32]: Is it possible to return theta as well ? For when we want to define the truth

Comment (Justinezgh) [reply sbids/tasks/utils.py:32]: or maybe I can just run 
```python 
trace(seed(lv, key)).get_trace()['theta']['value']
```
with the same seed as the one used for this 
``` python 
get_samples_and_scores(lv, key, batch_size=1)
```

Comment (Justinezgh) [reply sbids/tasks/utils.py:32]: hmm yes no sorry it's already in sample ^^

**Comment by Justinezgh** [comment]
> 
(the syntax is 

```python 
(log_probs, samples), scores = get_samples_and_scores(lotka_volterra, seed, batch_size=64)
```
)

**Comment by Justinezgh** [comment]
I tried to run the mcmc the same way you did in your notebook but it doesn't work, I end up with the same value 
image
[notebook](https://colab.research.google.com/drive/1-Pfc-O_5-Axvd7jgooCXsr77nhxkFIjy?usp=sharing)

**Comment by EiffL** [comment]
Hummmm, that's not great. You could try to set the initializer for the chain to a different one:
https://num.pyro.ai/en/stable/_modules/numpyro/infer/initialization.html

**Comment by Justinezgh** [comment]
ok, thanks :)

File: sbids/tasks/utils.py (hunk 1 above)
Comment (EiffL) [reply sbids/tasks/utils.py:32]: so, all good here :-) ?

**Comment by EiffL** [comment]
so, did it work? anything else to modify?

**Comment by Justinezgh** [comment]
yes it works but I didn't get how to change the ts argument 
image

**Comment by EiffL** [comment]
ok, so this is only a python question, what you want to do is to change the default value of a function parameter:
```python
from functools import partial
from sbids.tasks.lotkavolterra import lokta_volterra

lokta_volterra_10 = partial(lokta_volterra, ts=jnp.linspace(0,10,5))
```

**Comment by EiffL** [comment]
but the thing that blows my mind... is that on colab, on cpu, I get 6000 posterior samples in 15s....
https://colab.research.google.com/drive/1sowZ2oQHnbTwlY09IG7Wr5OsBNkidecx?usp=sharing
image
and it's returning the same result as what I was getting on my own computer:
image
except it was taking 20 mins on my GPU :-|

**Comment by EiffL** [comment]
yeah...... it's way faster on CPU :-| for some reason...

**Comment by Justinezgh** [comment]
yes on gpu it's 15min

**Comment by EiffL** [comment]
15s on CPU, not sure if it's the ode solver that sucks so much, or the nuts sampler.... but anyway, it's not very important for us.

**Review by Justinezgh — APPROVED** [review]
